nupic/research/frameworks/dynamic_sparse/networks/utils.py

```diff
@@ -0,0 +1,307 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2019, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+from collections import OrderedDict
+
+import torch
+
+from .layers import DSConv2d, DSLinear, DynamicSparseBase, RandDSConv2d, SparseConv2d
+
+# -------------------------------------------------
+# General Utils - network mutators
+# -------------------------------------------------
+
+
+def swap_layers(sequential, layer_type_a, layer_type_b):
+    """
+    If 'layer_type_a' appears immediately before 'layer_type_2',
+    this function will swap their position in a new sequential.
+
+    :param sequential: torch.nn.Sequential
+    :param layer_type_a: type of first layer
+    :param layer_type_a: type of second layer
+    """
+
+    old_seq = dict(sequential.named_children())
+    names = list(old_seq.keys())
+    modules = list(old_seq.values())
+
+    # Make copy of sequence.
+    new_seq = list(old_seq.items())
+
+    # Edit copy in place.
+    layer_a = modules[0]
+    name_a = names[0]
+    for i, (name_b, layer_b) in enumerate(list(old_seq.items())[1:], 1):
+
+        if isinstance(layer_a, layer_type_a) and isinstance(layer_b, layer_type_b):
+            new_seq[i - 1] = (name_b, layer_b)
+            new_seq[i] = (name_a, layer_a)
+
+        layer_a = layer_b
+        name_a = name_b
+
+    # Turn sequence into nn.Sequential.
+    new_seq = OrderedDict(new_seq)
+    new_seq = torch.nn.Sequential(new_seq)
+    return new_seq
+
+
+def squash_layers(sequential, *types, transfer_forward_hook=True):
+    """
+    This function squashes layers matching the sequence of 'types'.
+    For instance, if 'types' is [Conv2d, BatchNorm, KWinners] and
+    "sequential" has layers [..., Conv2d, BatchNorm, KWinners, ...],
+    then a new "sequential" will be returns of the form
+    [..., SubSequence, ...] where SubSequence calls .
```

Comment:
Sentence is incomplete here.  What does SubSequence call?

Reply:
Edited